File: task19/main.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <climits>

void writeFile(const std::string& filename, const std::vector<int>& data) {
    std::ofstream f(filename);
    for (int i = 0; i < (int)data.size(); ++i) {
        if (i > 0) f << ' ';
        f << data[i];
    }
}

std::vector<int> readFile(const std::string& filename) {
    std::ifstream f(filename);
    std::vector<int> data;
    int x;
    while (f >> x) data.push_back(x);
    return data;
}

void splitToFiles(const std::string& src,
                  const std::string& fileB,
                  const std::string& fileC,
                  int step) {
    std::vector<int> data = readFile(src);
    int n = (int)data.size();
    std::vector<int> bData, cData;
    int i = 0;
    bool toggle = true;
    while (i < n) {
        int end = std::min(i + step, n);
        std::vector<int>& target = toggle ? bData : cData;
        for (int j = i; j < end; ++j)
            target.push_back(data[j]);
        toggle = !toggle;
        i += step;
    }
    writeFile(fileB, bData);
    writeFile(fileC, cData);
}

void mergeFiles(const std::string& fileB,
                const std::string& fileC,
                const std::string& dst,
                int step) {
    std::vector<int> b = readFile(fileB);
    std::vector<int> c = readFile(fileC);
    std::vector<int> result;
    int ib = 0, ic = 0;
    int nb = (int)b.size(), nc = (int)c.size();

    while (ib < nb || ic < nc) {
        int bEnd = std::min(ib + step, nb);
        int cEnd = std::min(ic + step, nc);
        int bi = ib, ci = ic;

        while (bi < bEnd && ci < cEnd) {
            if (b[bi] <= c[ci]) result.push_back(b[bi++]);
            else                 result.push_back(c[ci++]);
        }
        while (bi < bEnd) result.push_back(b[bi++]);
        while (ci < cEnd) result.push_back(c[ci++]);

        ib += step;
        ic += step;
    }
    writeFile(dst, result);
}

void straightMergeSortExternal(const std::string& inputFile,
                               const std::string& outputFile) {
    std::vector<int> data = readFile(inputFile);
    int n = (int)data.size();
    writeFile(outputFile, data);

    const std::string fileB = "temp_b.txt";
    const std::string fileC = "temp_c.txt";

    for (int step = 1; step < n; step *= 2) {
        splitToFiles(outputFile, fileB, fileC, step);
        mergeFiles(fileB, fileC, outputFile, step);
    }

    std::remove(fileB.c_str());
    std::remove(fileC.c_str());
}

int main() {
    std::ofstream in("input.txt");
    in << "5 3 8 1 9 2 7 4 6";
    in.close();

    straightMergeSortExternal("input.txt", "output.txt");

    std::vector<int> result = readFile("output.txt");
    std::cout << "Результат: ";
    for (int x : result) std::cout << x << ' ';
    std::cout << std::endl;
    return 0;
}

File: task19/main.py
import os

def write_sequence(filename, data):
    with open(filename, 'w') as f:
        f.write(' '.join(map(str, data)))

def read_sequence(filename):
    with open(filename, 'r') as f:
        content = f.read().strip()
        if not content:
            return []
        return list(map(int, content.split()))

def split_to_files(src, file_b, file_c, step):
    data = read_sequence(src)
    n = len(data)
    b_data, c_data = [], []
    i = 0
    toggle = True
    while i < n:
        chunk = data[i:i + step]
        if toggle:
            b_data.extend(chunk)
        else:
            c_data.extend(chunk)
        toggle = not toggle
        i += step
    write_sequence(file_b, b_data)
    write_sequence(file_c, c_data)

def merge_files(file_b, file_c, dst, step):
    b = read_sequence(file_b)
    c = read_sequence(file_c)
    result = []
    ib, ic = 0, 0
    while ib < len(b) or ic < len(c):
        b_chunk = b[ib:ib + step]
        c_chunk = c[ic:ic + step]
        bi, ci = 0, 0
        while bi < len(b_chunk) and ci < len(c_chunk):
            if b_chunk[bi] <= c_chunk[ci]:
                result.append(b_chunk[bi]); bi += 1
            else:
                result.append(c_chunk[ci]); ci += 1
        result.extend(b_chunk[bi:])
        result.extend(c_chunk[ci:])
        ib += step
        ic += step
    write_sequence(dst, result)

def straight_merge_sort_external(input_file, output_file):
    data = read_sequence(input_file)
    n = len(data)
    write_sequence(output_file, data)

    file_b = 'temp_b.txt'
    file_c = 'temp_c.txt'
    step = 1

    while step < n:
        split_to_files(output_file, file_b, file_c, step)
        merge_files(file_b, file_c, output_file, step)
        step *= 2

    for f in [file_b, file_c]:
        if os.path.exists(f):
            os.remove(f)

if __name__ == '__main__':
    # Ручной ввод последовательности
    raw = input("Введите числа через пробел: ")
    data = list(map(int, raw.split()))

    # Вариант чтения из файла:
    # data = read_sequence('input.txt')

    write_sequence('input.txt', data)
    straight_merge_sort_external('input.txt', 'output.txt')
    print("Результат:", read_sequence('output.txt'))